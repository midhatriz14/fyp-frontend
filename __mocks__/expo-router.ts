const mockRouter = {
  push: jest.fn(),
};

export const useRouter = () => mockRouter;