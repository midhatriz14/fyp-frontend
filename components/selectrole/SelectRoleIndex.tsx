import { saveSecureData } from '@/store';
import { Picker } from '@react-native-picker/picker';
import { router } from 'expo-router';
import React, { useState } from 'react';
import { View, Text, TouchableOpacity, StyleSheet, Modal } from 'react-native';

export default function SelectRoleScreen() {
  const [modalVisible, setModalVisible] = useState(false);

  const toggleModal = () => {
    setModalVisible(!modalVisible);
  };
  const [selectedRole, setSelectedRole] = useState('');

  const handleConfirm = async () => {
    console.log('Selected Role:', selectedRole);
    await saveSecureData("role", selectedRole);
  };

  function handleOrganiserConfirm() {
    // throw new Error('Function not implemented.');
  }

  return (
    <View style={styles.container}>
      {/* Title */}
      <Text style={styles.title}>Select Your Role</Text>

      {/* Role Picker */}
      <Text style={styles.label}>Join As</Text>
      <TouchableOpacity style={styles.selectButton} onPress={toggleModal}>
        <Text style={styles.buttonText}>
          {selectedRole ? selectedRole : "Select"}
        </Text>
      </TouchableOpacity>

      <Modal
        visible={modalVisible}
        animationType="slide" // Slide animation
        transparent={false} // Ensure it's full-screen
      >
        <View style={styles.modalContainer}>
          <Picker
            testID="picker" // Add this testID
            selectedValue={selectedRole}
            onValueChange={(itemValue) => {
              setSelectedRole(itemValue);
              toggleModal();
            }}
            mode="dialog"
            style={styles.picker}
          >
            <Picker.Item label="Select" value="" />
            <Picker.Item label="Vendor" value="Vendor" />
            <Picker.Item label="Organizer" value="Organizer" />
          </Picker>
        </View>
      </Modal>

      {/* Confirm Button */}
      <TouchableOpacity
        testID="confirm-button" // Add this testID
        style={styles.confirmButton}
        onPress={() => {
          if (selectedRole === "Vendor") {
            handleConfirm();
            router.push("/bussinessselection");
          } else if (selectedRole === "Organizer") {
            handleConfirm();
            router.push("/signup");
          } else {
            alert("Please select a valid role!"); // Fallback if something goes wrong
          }
        }}
        disabled={!selectedRole} // Disable button if no role is selected
      >
        <Text style={styles.confirmButtonText}>Confirm</Text>
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#F7ECF5', // Light pink background
    paddingHorizontal: 20,
    paddingTop: 150,
    alignItems: 'center',
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    color: '#333',
    marginBottom: 20,
  },
  label: {
    fontSize: 16,
    color: '#333',
    alignSelf: 'flex-start',
    marginBottom: 10,
    paddingHorizontal: 30,
    marginTop: 50
  },
  confirmButton: {
    backgroundColor: '#780C60', // Dark purple color
    width: '80%',
    paddingVertical: 15,
    borderRadius: 10,
    alignItems: 'center',
    marginTop: 40,
  },
  confirmButtonText: {
    color: '#FFF',
    fontSize: 16,
    fontWeight: '600',
  },
  selectButton: {
    width: '80%',
    paddingVertical: 15,
    borderRadius: 10,
    borderWidth: 1,
    borderColor: '#CCC',
    backgroundColor: '#FFF',
    alignItems: 'center',
  },
  buttonText: {
    fontSize: 16,
    color: '#333',
  },
  modalContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  picker: {
    width: '80%',
    height: 200,
  },
});


