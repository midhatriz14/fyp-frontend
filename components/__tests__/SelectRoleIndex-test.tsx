import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import SelectRoleScreen from "../selectrole/SelectRoleIndex";
import { router } from "expo-router";
import SelectRoleIndex from "../selectrole/SelectRoleIndex";

jest.mock("@/store", () => ({
  saveSecureData: jest.fn(),
}));

jest.mock("expo-router", () => ({
  router: {
    push: jest.fn(),
  },
}));
// Mock global alert
global.alert = jest.fn();

describe("SelectRoleScreen", () => {
  // Clean up mocks after each test
  afterEach(() => {
    jest.clearAllMocks(); // Reset all mock data
  });

  it("renders correctly", () => {
    const { getByText } = render(<SelectRoleScreen />);
    expect(getByText("Select Your Role")).toBeTruthy();
    expect(getByText("Join As")).toBeTruthy();
    expect(getByText("Select")).toBeTruthy();
  });

  it("has the correct default state", () => {
    const { getByText } = render(<SelectRoleScreen />);
    expect(getByText("Select")).toBeTruthy(); // Default role is 'Select'
  });

  it("updates the state when a role is selected", () => {
    const { getByText, getByTestId } = render(<SelectRoleScreen />);
    fireEvent.press(getByText("Select")); // Open modal
    fireEvent(getByTestId("picker"), "onValueChange", "Vendor"); // Select "Vendor"
    expect(getByText("Vendor")).toBeTruthy(); // Confirm state update
  });

  it("navigates to the correct route for Vendor", async () => {
    const { getByText, getByTestId } = render(<SelectRoleScreen />);
    fireEvent.press(getByText("Select")); // Open modal
    fireEvent(getByTestId("picker"), "onValueChange", "Vendor"); // Select "Vendor"
    fireEvent.press(getByText("Confirm")); // Confirm selection
    await waitFor(() => {
      expect(router.push).toHaveBeenCalledWith("/bussinessselection");
    });
  });

  it("navigates to the correct route for Organizer", async () => {
    const { getByText, getByTestId } = render(<SelectRoleScreen />);
    fireEvent.press(getByText("Select")); // Open modal
    fireEvent(getByTestId("picker"), "onValueChange", "Organizer"); // Select "Organizer"
    fireEvent.press(getByText("Confirm")); // Confirm selection
    await waitFor(() => {
      expect(router.push).toHaveBeenCalledWith("/signup");
    });
  });

  it("displays an alert if no role is selected", () => {
    global.alert = jest.fn(); // Mock the alert function

    const { getByText } = render(<SelectRoleScreen />);

    // Simulate pressing the Confirm button
    fireEvent.press(getByText("Confirm"));

    // Check that the alert was triggered
    expect(global.alert).toHaveBeenCalledWith("Please select a valid role!");
  });

  it("enables the Confirm button when a role is selected", () => {
    const { getByTestId, getByText } = render(<SelectRoleScreen />);

    // Simulate selecting a role
    fireEvent.press(getByText("Select")); // Open the picker modal
    fireEvent(getByTestId("picker"), "onValueChange", "Vendor"); // Select "Vendor"

    // Find the Confirm button using the testID
    const confirmButton = getByTestId("confirm-button");

    // Check that the Confirm button is enabled
    expect(confirmButton.props.accessibilityState.disabled).toBe(false);
  });
});
