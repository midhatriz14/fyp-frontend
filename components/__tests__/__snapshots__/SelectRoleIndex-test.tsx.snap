// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectRoleScreen Snapshot Tests matches snapshot after confirming a selection 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F7ECF5",
      "flex": 1,
      "paddingHorizontal": 20,
      "paddingTop": 150,
    }
  }
>
  <Text
    style={
      {
        "color": "#333",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 20,
      }
    }
  >
    Select Your Role
  </Text>
  <Text
    style={
      {
        "alignSelf": "flex-start",
        "color": "#333",
        "fontSize": 16,
        "marginBottom": 10,
        "marginTop": 50,
        "paddingHorizontal": 30,
      }
    }
  >
    Join As
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFF",
        "borderColor": "#CCC",
        "borderRadius": 10,
        "borderWidth": 1,
        "opacity": 1,
        "paddingVertical": 15,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#333",
          "fontSize": 16,
        }
      }
    >
      Select
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#780C60",
        "borderRadius": 10,
        "marginTop": 40,
        "opacity": 1,
        "paddingVertical": 15,
        "width": "80%",
      }
    }
    testID="confirm-button"
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontSize": 16,
          "fontWeight": "600",
        }
      }
    >
      Confirm
    </Text>
  </View>
</View>
`;

exports[`SelectRoleScreen Snapshot Tests matches snapshot after selecting a role 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F7ECF5",
      "flex": 1,
      "paddingHorizontal": 20,
      "paddingTop": 150,
    }
  }
>
  <Text
    style={
      {
        "color": "#333",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 20,
      }
    }
  >
    Select Your Role
  </Text>
  <Text
    style={
      {
        "alignSelf": "flex-start",
        "color": "#333",
        "fontSize": 16,
        "marginBottom": 10,
        "marginTop": 50,
        "paddingHorizontal": 30,
      }
    }
  >
    Join As
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFF",
        "borderColor": "#CCC",
        "borderRadius": 10,
        "borderWidth": 1,
        "opacity": 1,
        "paddingVertical": 15,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#333",
          "fontSize": 16,
        }
      }
    >
      Select
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#780C60",
        "borderRadius": 10,
        "marginTop": 40,
        "opacity": 1,
        "paddingVertical": 15,
        "width": "80%",
      }
    }
    testID="confirm-button"
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontSize": 16,
          "fontWeight": "600",
        }
      }
    >
      Confirm
    </Text>
  </View>
</View>
`;

exports[`SelectRoleScreen Snapshot Tests matches snapshot when modal is open 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F7ECF5",
      "flex": 1,
      "paddingHorizontal": 20,
      "paddingTop": 150,
    }
  }
>
  <Text
    style={
      {
        "color": "#333",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 20,
      }
    }
  >
    Select Your Role
  </Text>
  <Text
    style={
      {
        "alignSelf": "flex-start",
        "color": "#333",
        "fontSize": 16,
        "marginBottom": 10,
        "marginTop": 50,
        "paddingHorizontal": 30,
      }
    }
  >
    Join As
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFF",
        "borderColor": "#CCC",
        "borderRadius": 10,
        "borderWidth": 1,
        "opacity": 1,
        "paddingVertical": 15,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#333",
          "fontSize": 16,
        }
      }
    >
      Select
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#780C60",
        "borderRadius": 10,
        "marginTop": 40,
        "opacity": 1,
        "paddingVertical": 15,
        "width": "80%",
      }
    }
    testID="confirm-button"
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontSize": 16,
          "fontWeight": "600",
        }
      }
    >
      Confirm
    </Text>
  </View>
</View>
`;

exports[`SelectRoleScreen Snapshot Tests matches the initial render snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F7ECF5",
      "flex": 1,
      "paddingHorizontal": 20,
      "paddingTop": 150,
    }
  }
>
  <Text
    style={
      {
        "color": "#333",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 20,
      }
    }
  >
    Select Your Role
  </Text>
  <Text
    style={
      {
        "alignSelf": "flex-start",
        "color": "#333",
        "fontSize": 16,
        "marginBottom": 10,
        "marginTop": 50,
        "paddingHorizontal": 30,
      }
    }
  >
    Join As
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFF",
        "borderColor": "#CCC",
        "borderRadius": 10,
        "borderWidth": 1,
        "opacity": 1,
        "paddingVertical": 15,
        "width": "80%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#333",
          "fontSize": 16,
        }
      }
    >
      Select
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#780C60",
        "borderRadius": 10,
        "marginTop": 40,
        "opacity": 1,
        "paddingVertical": 15,
        "width": "80%",
      }
    }
    testID="confirm-button"
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontSize": 16,
          "fontWeight": "600",
        }
      }
    >
      Confirm
    </Text>
  </View>
</View>
`;
